src/gradients.rs

```diff
@@ -101,6 +101,62 @@ impl Tape for NoneTape {
     fn add_backward_op<F: 'static + FnOnce(&mut Gradients)>(&mut self, _operation: F) {}
 }
 
+/// An error that indicates the parameters at
+/// [UnusedParamsError::param_locations] were not used during gradient computation,
+/// and therefore don't have gradients associated with them.
+#[derive(Debug)]
+pub struct UnusedParamsError {
+    pub param_locations: Vec<String>,
+}
+
+impl UnusedParamsError {
+    pub fn one() -> Self {
+        Self {
+            param_locations: vec!["".into()],
+        }
+    }
+
+    pub fn prepend(mut self, location: &str) -> Self {
+        for p in self.param_locations.iter_mut() {
+            p.insert_str(0, location);
+        }
+        self
+    }
+}
+
+pub trait CollectUnused<T> {
+    /// A method solely for adding two `Result<(), UnusedParamsError>` together.
+    ///
+    /// Cases:
+    /// 1. `self` and `b` are `Ok(())`; does nothing
+    /// 2. `self` is `Ok(())`, `b` is `Err(_)`; `*self = b`
+    /// 3. `self` is `Err(_)`, `b` is `Ok(())`; does nothing
+    /// 4. `self` and `b` are `Err(_)`; adds the unused params from `b` into `self`.
+    fn maybe_add_unused(&mut self, tag: &str, b: Self);
```

Comment:
Still feel like this could be named better, not exactly sure what to call it. will brainstorm a bit